File: server/app/schemas/books.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class BookMetadataBase(BaseModel):
    """Base model for book metadata."""

    title: str
    author: str
    description: str = ""
    cover_url: str | None = None
    format: str  # "PDF" or "EPUB"
    file_url: str
    file_size_bytes: int | None = Field(None, le=5 * 1024 * 1024)  # 5MB max
    num_pages: int | None = None
    pdf_toc: dict[str, Any] | list[dict[str, Any]] | None = None  # Allow both dict and list
    epub_chapter_char_counts: list[int] | None = None
    epub_page_char_counts: list[int] | None = None

    @field_validator("format")
    @classmethod
    def validate_format(cls, v: str) -> str:
        if v not in ["PDF", "EPUB"]:
            raise ValueError("Format must be either 'PDF' or 'EPUB'")
        return v


class BookMetadataCreate(BookMetadataBase):
    """Model for creating book metadata."""

    pass


class BookMetadataUpdate(BaseModel):
    """Model for updating book metadata."""

    title: str | None = None
    author: str | None = None
    description: str | None = None
    cover_url: str | None = None
    file_url: str | None = None
    file_size_bytes: int | None = None
    num_pages: int | None = None
    pdf_toc: dict[str, Any] | list[dict[str, Any]] | None = None  # Allow both dict and list
    epub_chapter_char_counts: list[int] | None = None


class BookMetadataResponse(BookMetadataBase):
    """Model for book metadata response."""

    id: UUID
    # Allow created_at to be either string or datetime
    created_at: str | datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator("created_at", mode="before")
    @classmethod
    def validate_datetime(cls, value: str | datetime) -> str:
        """Convert datetime to string if it's a datetime object."""
        if isinstance(value, datetime):
            return value.isoformat()
        return value


class UserBookLibraryBase(BaseModel):
    """Base model for user book library."""

    user_id: UUID
    book_metadata_id: UUID
    pdf_current_page: int | None = None
    epub_progress: dict[str, Any] | None = None


class UserBookLibraryCreate(UserBookLibraryBase):
    """Model for creating user book library."""

    pass


class UserBookLibraryUpdate(BaseModel):
    """Model for updating user book library."""

    pdf_current_page: int | None = None
    epub_progress: dict[str, Any] | None = None


class UserBookLibraryResponse(UserBookLibraryBase):
    """Model for user book library response."""

    id: UUID
    # Allow date_added to be either string or datetime
    date_added: str | datetime
    book_metadata: BookMetadataResponse

    model_config = ConfigDict(from_attributes=True)

    @field_validator("date_added", mode="before")
    @classmethod
    def validate_date_added(cls, value: str | datetime) -> str:
        """Convert datetime to string if it's a datetime object."""
        if isinstance(value, datetime):
            return value.isoformat()
        return value

File: packages/shared/src/api/types/books.ts
// Book types aligned with backend schema and database
export type BookFormat = "PDF" | "EPUB";
export type HighlightColor = "YELLOW" | "GREEN" | "BLUE" | "PINK" | "PURPLE";

// Base types matching database schema
export type BookMetadata = {
  id: string;
  title: string;
  author: string | null;
  description: string | null;
  cover_url: string | null; // Matches backend and database field name
  format: BookFormat;
  file_url: string; // Matches backend and database field name
  file_size_bytes: number | null; // Matches backend and database field name
  num_pages: number | null; // Matches backend and database field name
  pdf_toc: unknown | null; // JSON type from database
  epub_chapter_char_counts: number[] | null; // Matches backend and database
  epub_page_char_counts: number[] | null; // Matches backend and database
  created_at: string;
};

export type UserBookLibrary = {
  id: string;
  user_id: string;
  book_metadata_id: string;
  pdf_current_page: number | null;
  epub_progress: unknown | null; // JSON type from database
  date_added: string;
  book_metadata: BookMetadata;
};

export type Highlight = {
  id: string;
  user_book_lib_id: string;
  original_text: string;
  color: HighlightColor;
  note?: string | null;
  html_range?: SerializedRangeJson | null;
  chapter_idx?: number | null;
  chapter_href?: string | null;
  chapter_title?: string | null;
  page?: number | null;
  pdf_rect_position?: Record<string, unknown> | null;
};

// Create types
export type BookMetadataCreate = Omit<
  BookMetadata,
  "id" | "created_at" | "updated_at"
>;
export type UserBookLibraryCreate = Omit<
  UserBookLibrary,
  "id" | "date_added" | "book_metadata"
>;
export type HighlightCreate = Omit<Highlight, "id">;

// Update types
export type BookMetadataUpdate = Partial<
  Omit<BookMetadata, "id" | "created_at" | "updated_at">
>;
export type UserBookLibraryUpdate = Partial<
  Pick<UserBookLibrary, "epub_progress" | "pdf_current_page">
>;
export type HighlightUpdate = Partial<Pick<Highlight, "color" | "note">>;

// Progress types
export interface EpubProgress {
  globalProgress: {
    current: number;
    total: number;
  };
  loc?: string;
}

// Type guard for EpubProgress from JSON
export function isEpubProgress(progress: unknown): progress is EpubProgress {
  return (
    progress !== null &&
    typeof progress === "object" &&
    "globalProgress" in progress &&
    typeof progress.globalProgress === "object" &&
    progress.globalProgress !== null &&
    "current" in progress.globalProgress &&
    "total" in progress.globalProgress &&
    typeof progress.globalProgress.current === "number" &&
    typeof progress.globalProgress.total === "number"
  );
}

// Extended types for frontend use
export type UserBookLibraryWithTypedProgress = Omit<
  UserBookLibrary,
  "epub_progress"
> & {
  epub_progress: EpubProgress | null;
};

export type BookViewProps = BookMetadata & {
  library_id?: string | null;
  pdf_current_page?: number | null;
  epub_progress?: EpubProgress | null;
};

// User Book Library API Types
export interface UserBookLibraryProgressResponse {
  id: string;
  user_id: string;
  book_metadata_id: string;
  pdf_current_page?: number;
  epub_progress?: EpubProgress;
  date_added: string;
  book_metadata: BookMetadata;
}

// Highlight API Types
export interface HighlightCreateRequest {
  user_book_lib_id: string;
  original_text: string;
  color: HighlightColor;
  note?: string | null;
  html_range?: SerializedRangeJson | null;
  chapter_idx?: number | null;
  chapter_href?: string | null;
  chapter_title?: string | null;
  page?: number | null;
  pdf_rect_position?: Record<string, unknown> | null;
}

export interface HighlightUpdateRequest {
  note: string;
  text: string;
}

// PDF.js Related Types
export interface PdfDocumentProxy {
  numPages: number;
  fingerprint: string;
  [key: string]: unknown;
}

// EPUB highlight range serialization types
export interface SerializedRange {
  startContainerPath: number[];
  startOffset: number;
  endContainerPath: number[];
  endOffset: number;
}

// Helper type for converting SerializedRange to JSON
export interface SerializedRangeJson {
  startContainerPath: number[];
  startOffset: number;
  endContainerPath: number[];
  endOffset: number;
}

// Type guard for SerializedRange
export function isSerializedRange(obj: unknown): obj is SerializedRange {
  return (
    obj !== null &&
    typeof obj === "object" &&
    "startContainerPath" in obj &&
    "startOffset" in obj &&
    "endContainerPath" in obj &&
    "endOffset" in obj &&
    Array.isArray((obj as unknown as SerializedRange).startContainerPath) &&
    Array.isArray((obj as unknown as SerializedRange).endContainerPath) &&
    typeof (obj as unknown as SerializedRange).startOffset === "number" &&
    typeof (obj as unknown as SerializedRange).endOffset === "number"
  );
}

// Convert SerializedRange to JSON-safe format
export function serializeRangeToJson(
  range: SerializedRange,
): SerializedRangeJson {
  return {
    startContainerPath: range.startContainerPath,
    startOffset: range.startOffset,
    endContainerPath: range.endContainerPath,
    endOffset: range.endOffset,
  };
}

// React-pdf-highlighter-extended types
export interface PdfHighlightPosition {
  boundingRect: {
    x1: number;
    y1: number;
    x2: number;
    y2: number;
    width: number;
    height: number;
  };
  rects: Array<{
    x1: number;
    y1: number;
    x2: number;
    y2: number;
    width: number;
    height: number;
  }>;
  pageNumber: number;
}

export interface PdfHighlightContent {
  text?: string;
  image?: string;
}

export interface PdfHighlight {
  id: string;
  position: PdfHighlightPosition;
  content: PdfHighlightContent;
  color?: string;
  note?: string;
}
